core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java

```diff
@@ -284,6 +283,11 @@ private void decPendingAndFinishIfNeeded() {
 
     private void finish() {
         if (finished.compareAndSet(false, true)) {
+            try {
+                primary.updateGlobalCheckpoint();
```

Comment:
Instead of actively calling this on every completed operation, I wish this was done automatically by GlobalCheckpointTracker whenever a local checkpoint is incremented. IndexShard could register a listener on GlobalCheckpointTracker that would be invoked whenever the global checkpoint increases due to a local checkpoint update and `globalCheckpointSyncer` could implement this listener and act accordingly.